README.md
A small summer project, to get better with React and its logic.

Some things to note:
 - This work is purely for self-educational purposes, it is not intended for distribution, hence the lack of proper testing, documentation and standard procedures.
 - I am still very much influenced by old Javascript design patterns, this is why I'm trying to create something more complex.
 - I will probably never finish this project but still keep it here if I will ever need it in the future.
 

CHANGELOG.md
# Changelog

## [Unreleased]

## [0.10.0] - 2017-08-1
### Added
 - Interaction with user
 - Count points (correct answers)

## [0.9.0] - 2017-07-31
### Added
 - Operations now work

## [0.8.0] - 2017-07-31
### Added
 - Implemented a test version for the display of operations

## [0.7.0] - 2017-07-25
### Added
 - Fraction semplification

## [0.6.0] - 2017-07-25
### Added
 - Division

## [0.5.0] - 2017-07-25
### Added
 - Multiplication

## [0.4.0] - 2017-07-24
### Added
 - Added pythagoremic tables

## [0.3.0] - 2017-07-18
### Added
 - Subtraction: negative and positive

## [0.2.0] - 2017-07-17
### Added
 - Addition

## [0.1.0] - 2017-07-15
### Added
 - Scheleton of the app


src/App.js
import React, { Component } from 'react';
import { Addition, Subtraction, Multiplication, Division, Pythagoras } from './Services.js';
import Exercise from './Exercise.js';
import Settings from './Settings.js'
import './App.css';


class App extends Component {
  constructor() {
    super()
    /*this.addition = new Addition(1);
    this.subtraction = new Subtraction(0.2, true);
    this.pythagoras = new Pythagoras({ exerciseLength: 5 })
    this.multiplication = new Multiplication(0.2)
    this.division = new Division(0.5, 2);*/
    this.state = {
      //0: addition, 1: subtraction, 2: multiplication, 3: division
      operation: "",
      mode: "",
      amount: "",
      difficulty: 0.5
    }
    this.updateSettings = this.updateSettings.bind(this);
    this.addOperation = this.addOperation.bind(this);
  }
  updateSettings(inputSettings) {
    console.log("updating State")

    this.setState(prevState => ({
      operation: inputSettings.operation || prevState.operation,
      mode: inputSettings.mdoe || prevState.mode,
      difficulty: inputSettings.difficulty || prevState.difficulty,
      amount: inputSettings.amount || prevState.amount,
      scene: inputSettings.scene || prevState.scene
    }))
  }
  addOperation(operationID, mode, amount) {
    this.setState(prevState => ({
      operation: prevState.operation+","+operationID,
      mode: prevState.mode+","+mode,
      amount: prevState.amount + "," + amount
    }))
  }
  render() {
    //let operation = this.prova.toArray();
    /*let operations = []
    for (let i = 0; i < 1; i++) {
      operations[i] = this.subtraction.toArray();
    }
    let operatore = this.addition.operatore
    let sub = this.addition.toArray()

    let exercise = this.pythagoras.newExercise();

    let mul = this.multiplication.toArray();
    let div = this.division.toArray();*/
    return (
      <div className="App">
        <Settings settings={this.state} updateSettings={this.updateSettings} addOperation={this.addOperation}/>
        <Exercise settings={this.state} eventHandler={this.updateSettings} />

        {/*operations.map(function (operation, u) {
          return (
            <div className="operation sum-sub" key={u}>
              <div className="topPart">
                <div className="operandi">
                  <span className="firstNumber">{operation[0]}</span>
                  <span className="secondNumber">{operation[1]}</span>
                </div>
                <div className="operatore">{operatore}</div>
              </div>
              <div className="bottomPart">
                <div className="result">{operation[2]}</div>
              </div>
            </div>
          )
        })
        <div className="operation sum-sub">
          <div className="topPart">
            <div className="operandi">
              <span className="firstNumber">{sub[0]}</span>
              <span className="secondNumber">{sub[1]}</span>
            </div>
            <div className="operatore">{this.subtraction.operatore}</div>
          </div>
          <div className="bottomPart">
            <div className="result">{sub[2]}</div>
          </div>
        </div>
        <div className="operation sum-sub">
          <div className="topPart">
            <div className="operandi">
              <span className="firstNumber">{mul[0]}</span>
              <span className="secondNumber">{mul[1]}</span>
            </div>
            <div className="operatore">{this.multiplication.operatore}</div>
          </div>
          <div className="bottomPart">
            <div className="result">{mul[2]}</div>
          </div>
        </div>
        <div className="operation sum-sub">
          <div className="topPart">
            <div className="operandi">
              <span className="firstNumber">{div[0]}</span>
              <span className="secondNumber">{div[1]}</span>
            </div>
            <div className="operatore">{this.division.operatore}</div>
          </div>
          <div className="bottomPart">
            <div className="result">{div[2][0]} / {div[2][1]}</div>
          </div>
        </div>*/}
        {/*exercise.map(function (operation, u) {
          return (
            <div className="operation sum-sub" key={u}>
              <div className="topPart">
                <div className="operandi">
                  <span className="firstNumber">{operation[0]}</span>
                  <span className="secondNumber">{operation[1]}</span>
                </div>
                <div className="operatore">X</div>
              </div>
              <div className="bottomPart">
                <div className="result">{operation[2]}</div>
              </div>
            </div>
          )
        })*/}
      </div>
    );
  }
}


export default App;


src/Exercise.js
import React, { Component } from 'react';
import { Addition, Subtraction, Multiplication, Division } from './Services.js';
/*
* TODO
*  There should be a list of all actual operations
*  but not generated here, here there should only be
*  a list of the numbers and results, the app should
*  give to Exercise the actual list of operations, so
*  that it an be scrambled properly and so that there
*  can be metadata attached, like the fact that it's
*  a right or a wrong answer and if it's visible or not
*
*  Every operation should not be visible, if not for
*  the only one which is being asked right now
*  so the scanning through all the array can be
*  reduced to scanning through only the operations up
*  to the one which is displayed
*/
class Operation extends Component {
  constructor() {
    super()
    this.state = {
      mode: "hidden",
      left: -1
    }
  }
  render() {
    let operations = this.props.operations
    return (
      {/*
      <div className="operation sum-sub" key={u}>
        <div className="topPart">
          <div className="operandi">
            <span className="firstNumber">{operation[0]}</span>
            <span className="secondNumber">{operation[1]}</span>
          </div>
          <div className="operatore">X</div>
        </div>
        <div className="bottomPart">
          <div className="result">{operation[2]}</div>
        </div>
      </div>*/}
    )
  }
}


class Exercise extends Component {
  constructor() {
    super();
    console.log("before constructor")
    this.operation = {}
    this.startExercise = this.startExercise.bind(this);
    this.state = {
      visible: 0,
      points: 0
    }
    this.timer = 0
    console.log("after constructor")
    this.submitResult = this.submitResult.bind(this);
  }

  startExercise(e, elements) {
    //console.log("started new exercise")
    //console.log(e.target)
    /*let visibilities = []
    visibilities[0] = true
    for (let i = 1; i < elements.length; i++) {
      visibilities[i] = false
    }*/

    this.setState(prevState => ({
      scene: "playing",
      points: 0,
      visible: 0
    }));
    this.timer = Date.now()
    //this.props.eventHandler(e);
  }
  answerOperation(e) {

  }
  componentWillUpdate() {

  }
  componentDidUpdate() {
    console.log(this.props)
  }
  parseOperation(settings) {
    let operationsID = settings.operation.split(',');
    let modes = settings.mode.split(',');
    let amounts = settings.amount.split(',');
    let operations = []
    for (let i = 0; i < operationsID.length; i++) {
      let mode = parseInt(modes[i], 10)
      let amount = parseInt(amounts[i], 10)
      switch (parseInt(operationsID[i], 10)) {
        case 0:
          operations[i] = new Addition(settings.difficulty);
          break;
        case 1:
          operations[i] = new Subtraction(settings.difficulty, mode);
          break;
        case 2:
          operations[i] = new Multiplication(settings.difficulty);
          break;
        case 3:
          operations[i] = new Division(settings.difficulty, mode);
          break;
        default:
          operations[i] = {};
      }
    }
    return operations
  }
  submitResult(e, result, element) {
    console.log(e.target.value)
    let point = (result===parseInt(e.target.value,10))? 1 : 0
    this.setState(prevState => {
      console.log(element)
      /*let visibilities = prevState.visibility;
      visibilities[element] = false
      if (element + 1 < visibilities.length) {
        visibilities[element + 1] = true;
      }
      console.log(prevState.points)*/
    
      return ({
        points: prevState.points + point,
        visible: element+1
      })
    })
    return true;
  }

  componentDidUpdate() {
    if(this.resultInput)
      this.resultInput.focus();   
  }
  render() {
    let settings = this.props.settings;
    this.operations = this.parseOperation(settings)
    this.amounts = this.props.settings.amount.split(',').map((amount) => {
      return parseInt(amount, 10);
    });
    this.elements = []
    let j = 0;
    let e = 0;

    for (let a in this.operations) {
      for (let i = 0; i < this.amounts[a]; i++) {
        this.elements[e++] = [this.operations[a].toArray(), this.operations[a].operatore];
      }
    }

    let self = this;

    console.log(settings)
    console.log(this.operations)
    return (
      <div className="exercise">
        {this.state.points}
        <div onClick={(e) => {
          this.startExercise(e, self.elements)
        }}> Start</div>
        {this.elements.map(function (operation, u, array) {
          console.log(self.state.visible)
          if (self.state.visible === u || self.state.visible === u-1  || self.state.visible === u+1) {
            return (
              <div className="prova" key={u}>
                <div className="operation sum-sub" key={u}>
                  <div className="topPart">
                    <div className="operandi">
                      <span className="firstNumber">{operation[0][0]}</span>
                      <span className="secondNumber">{operation[0][1]}</span>
                    </div>
                    <div className="operatore">{operation[1]}</div>
                  </div>

                  <div className="bottomPart">
                    <input onKeyPress={(e) => {
                      if (e.key == 'Enter') {
                        self.submitResult(e, operation[0][2], u)
                      }  
                    }} className="result" ref={(input) => { if (self.state.visible === u) { self.resultInput = input; } }} />
                  </div>
                </div>
              </div>
            )
          } else {
            return ("");
          }
        })}
      </div>
    );
  }
}

export default Exercise;

src/Services.js
//const shuffleArray = (arr) => arr.sort(() => (Math.random() - 0.5))
class Exercise {
  getRandomInt(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    if (min === max)
      return max;

    return Math.floor(Math.random() * (max + 1 - min)) + min;
  }
}
class Pythagoras extends Exercise {
  constructor(exerciseLength) {
    super();
    this.numbers = []
    let i = 0
    for (let a = 2; a < 13; a++) {
      for (let b = a; b < 13; b++) {
        this.numbers[i++] = [a, b, a * b]
      }
    }
    this.cacheSize = 20;
    this.cache = [];

    this.exerciseLength = exerciseLength
    /*this.useCache = false
    if (this.exerciseLength > this.numbers.length) {
      this.useCache = true
    }*/
  }

  newOperation() {
    if (this.cache.length === this.cacheSize/* && this.useCache*/) {
      this.numbers.push(this.cache.pop())
    }

    let index = super.getRandomInt(0, this.numbers.length - 1)
    let numbers = this.numbers[index]

    //if (this.useCache) {
      this.cache.push(numbers)
    //}

    this.numbers.splice(index, 1);
    return numbers;
  }
  toArray() {
    return this.newOperation();
  }

  newExercise() {
    let exercise = []
    for (let a = 0; a < this.exerciseLength; a++) {
      exercise[a] = this.newOperation();
    }
    return exercise;
  }
}
class Operation {
  getRandomInt(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    if (min === max)
      return max;

    return Math.floor(Math.random() * (max + 1 - min)) + min;
  }
  getRandomDigit() {
    return Math.floor(Math.random() * 10);
  }
  coinFlip() {
    return (Math.floor(Math.random() * 2) === 0);
  }
  toNumber(arr) {
    if (typeof arr === "number") {
      return arr
    } else {
      return parseInt(arr.slice(0).reverse().join(''), 10);
    }  
  }
  toArray(number) {
    let string = '' + number
    return string.split('').slice(0).reverse()
  }
  //https://jsfiddle.net/JamesOR/RC7SY/
  getAllFactorsFor(remainder) {
    var factors = [], i;

    for (i = 2; i <= remainder; i++) {
      while ((remainder % i) === 0) {
        factors.push(i);
        let temp = []
        for (let a = 0; a < factors.length - 1; a++) {
          temp[a] = factors[factors.length - 1] * factors[a];
        }
        console.log(temp)
        factors = factors.concat(temp)
        remainder /= i;
      }
    }

    return factors;
  }
  //https://stackoverflow.com/questions/28575416/finding-out-if-two-numbers-are-relatively-prime
  gcd(a, b) {
    let t;
    while (b !== 0) {
      t = a;
      a = b;
      b = t % b;
    }
    return a;
  }
  relativelyPrime(a, b) {
    return this.gcd(a, b) === 1;
  }
}
class Addition extends Operation {
  //Difficulty goes from 0 to 1.
  constructor(difficulty) {
    super();
    this.difficulty = Math.min(Math.max(difficulty, 0.2), 1);

    this.maxLen = 10
    this.maxRiporti = 5
    this.maxDifficulty = this.maxLen + this.maxRiporti
  }
  get operatore() {
    return "+";
  }
  generateOperation() {
    //D is the effective difficulty (it's not descriptive but doesn't waste space)
    let D = Math.round(this.difficulty * this.maxDifficulty);
    //The number of carries in the operation 
    let carry = super.getRandomInt(1, this.maxRiporti * this.difficulty);
    //The number of digits in the number
    let nDigit = Math.max(D - carry, 3)

    //If there arent enough digits
    if (nDigit + carry < D) {
      nDigit = D - carry
    }

    //Number of digits in first number
    let nDigit0 = Math.floor(nDigit / 2)
    //Digits in secon number
    let nDigit1 = nDigit - nDigit0;

    let remain = Math.min(nDigit0, nDigit1)
    let firstNumber = []
    let secondNumber = []
    let Digit = super.getRandomDigit();
    let newDigit = -1;
    let carryLeft = carry;
    while (remain > 0) {
      //If Digit is 0 then I can't have a carry
      if (Digit === 0 && carryLeft >= 0) {
        Digit = super.getRandomInt(1, 9)
      }
      if (super.coinFlip() === true && (remain - carryLeft) > 0) {
        newDigit = super.getRandomInt(0, 10 - Digit - 1)
      } else {
        newDigit = super.getRandomInt(10 - Digit, 9)
        carryLeft--
      }
      firstNumber.push(Digit)
      secondNumber.push(newDigit)
      Digit = super.getRandomDigit();
      remain--;
    }

    let l = firstNumber.length
    for (let j = l; j < nDigit0; j++) {
      firstNumber[j] = super.getRandomInt(1, 9)
    }
    l = secondNumber.length
    for (let j = l; j < nDigit1; j++) {
      secondNumber[j] = super.getRandomInt(1, 9)
    }

    return [firstNumber, secondNumber]
  }
  toArray() {
    let numbers = this.generateOperation()
    let firstNumber = super.toNumber(numbers[0]);
    let secondNumber = super.toNumber(numbers[1]);
    let result = firstNumber + secondNumber;
    return [firstNumber, secondNumber, result]
  }
}
class Subtraction extends Operation {
  constructor(difficulty, negative) {
    super();
    this.difficulty = Math.min(Math.max(difficulty, 0.2), 1);
    this.negative = (negative === 0)? false : true;
    this.maxLen = 10
    this.maxRiporti = 5
    this.maxDifficulty = this.maxLen + this.maxRiporti
  }
  get operatore() {
    return "-";
  }
  generateNonNegativeOperation() {
    //D is the effective difficulty (it's not descriptive but doesn't waste space)
    let D = Math.round(this.difficulty * this.maxDifficulty);
    //The number of carries in the operation 
    let carry = super.getRandomInt(1, this.maxRiporti * this.difficulty);
    //The number of digits in the number
    let nDigit = Math.max(D - carry, 4)

    //If there arent enough digits
    if (nDigit + carry < D) {
      nDigit = D - carry
    }

    //Number of digits in first number
    let nDigit0 = Math.round(nDigit / 2)
    if (nDigit0 < nDigit / 2) {
      nDigit0 = nDigit0 + 1
    }
    //Digits in secon number
    let nDigit1 = nDigit - nDigit0;

    let remain = Math.min(nDigit0, nDigit1)
    let firstNumber = []
    let secondNumber = []
    let Digit = super.getRandomInt(1, 9);
    let newDigit = -1;
    let carryLeft = carry;
    console.log(carry)
    while (remain > 0) {
      //If Digit is 9 then I can't have a carry
      if (Digit === 9 && (carryLeft) >= 0) {
        Digit = super.getRandomInt(1, 8)
      }
      if (Digit === 0) {
        carryLeft--
      }
      if (super.coinFlip() === true && (remain - carryLeft) > 0) {
        newDigit = super.getRandomInt(0, Digit)
      } else {
        newDigit = super.getRandomInt(Digit + 1, 9)
        carryLeft--
      }
      firstNumber.push(Digit)
      secondNumber.push(newDigit)
      Digit = super.getRandomDigit();
      remain--;
    }

    let l = firstNumber.length
    for (let j = l; j < nDigit0; j++) {
      firstNumber[j] = super.getRandomInt(1, 9)
    }
    l = secondNumber.length
    console.log(l)
    for (let j = l; j < nDigit1; j++) {
      console.log(j)
      secondNumber[j] = super.getRandomInt(1, 9)
    }
    let result = (super.toNumber(firstNumber) - super.toNumber(secondNumber));
    if (result < 0) {
      firstNumber.push(super.getRandomInt(1, 9))
    }
    return [firstNumber, secondNumber]
  }
  generateOperation() {
    //D is the effective difficulty (it's not descriptive but doesn't waste space)
    let D = Math.round(this.difficulty * this.maxDifficulty);
    //The number of carries in the operation 
    let carry = super.getRandomInt(1, this.maxRiporti);
    //The number of digits in the number
    let nDigit = Math.max(D - carry, 3)

    //If there arent enough digits
    if (nDigit + carry < D) {
      nDigit = D - carry
    }

    //Number of digits in first number
    let nDigit0 = Math.floor(nDigit / 2)
    //Digits in secon number
    let nDigit1 = nDigit - nDigit0;

    let remain = Math.max(nDigit0, nDigit1)
    let firstNumber = []
    let secondNumber = []
    let Digit = super.getRandomDigit();
    let newDigit = -1;
    let carryLeft = carry;
    while (remain > 0) {
      //If Digit is 9 then I can't have a carry
      if (Digit === 9 && (remain - carryLeft) <= 0) {
        Digit = super.getRandomInt(0, 8)
      }
      if (super.coinFlip() === true && (remain - carryLeft) > 0) {
        newDigit = super.getRandomInt(0, Digit)
      } else {
        newDigit = super.getRandomInt(Digit + 1, 9)
        carryLeft--
      }
      firstNumber.push(Digit)
      secondNumber.push(newDigit)
      Digit = super.getRandomDigit();
      remain--;
    }

    let l = firstNumber.length
    for (let j = l; j < nDigit0; j++) {
      firstNumber[j] = super.getRandomInt(1, 9)
    }
    l = secondNumber.length
    for (let j = l; j < nDigit1; j++) {
      secondNumber[j] = super.getRandomInt(1, 9)
    }

    return [firstNumber, secondNumber]
  }
  toArray() {
    let numbers = [];
    if (!this.negative) {
      numbers = this.generateNonNegativeOperation()
    } else {
      numbers = this.generateOperation()
    }

    let firstNumber = super.toNumber(numbers[0]);
    let secondNumber = super.toNumber(numbers[1]);
    let result = firstNumber - secondNumber;

    return [firstNumber, secondNumber, result]
  }
}
class Multiplication extends Operation {
  //Difficulty goes from 0 to 1.
  constructor(difficulty, mode) {
    super();
    this.difficulty = Math.min(Math.max(difficulty, 0.2), 1);
    this.length1 = (this.difficulty >= 0.5) ? 2 : 1;
    this.length0 = Math.round((this.difficulty * 6) - 2)

    //Adding Pythagoras table
    this.mode = mode; //mode = 0 normal, mode = 1 Pythagoras
    this.numbers = []
    let i = 0
    for (let a = 2; a < 13; a++) {
      for (let b = a; b < 13; b++) {
        this.numbers[i++] = [a, b]
      }
    }
    this.cacheSize = 20;
    this.cache = [];

  }
  get operatore() {
    return "X";
  }
  generatePythagoras() {
    if (this.cache.length === this.cacheSize/* && this.useCache*/) {
      this.numbers.push(this.cache.pop())
    }

    let index = super.getRandomInt(0, this.numbers.length - 1)
    let numbers = this.numbers[index]

    //if (this.useCache) {
      this.cache.push(numbers)
    //}

    this.numbers.splice(index, 1);
    return numbers;
  }
  generateOperation() {

    let firstNumber = []
    let secondNumber = []

    firstNumber[0] = super.getRandomDigit()
    secondNumber[0] = super.getRandomDigit()

    for (let a = 1; a < this.length0; a++) {
      firstNumber[a] = super.getRandomInt(1, 9)
    }
    for (let a = 1; a < this.length1; a++) {
      secondNumber[a] = super.getRandomInt(1, 9)
    }

    return [firstNumber, secondNumber]
  }
  toArray() {

    let numbers = (this.mode === 0) ? this.generateOperation() : this.generatePythagoras();
    let firstNumber = super.toNumber(numbers[0]);
    let secondNumber = super.toNumber(numbers[1]);
    let result = firstNumber * secondNumber;
    return [firstNumber, secondNumber, result]
  }
}
class Division extends Operation {
  constructor(difficulty, mode) {
    super()
    //Mode:
    //0 : extact division
    //1 : approximate division
    this.mode = mode
    this.difficulty = this.difficulty = Math.min(Math.max(difficulty, 0.2), 1);


  }
  get operatore() {
    return "/"
  }
  generateFraction() {
    let firstNumber = super.getRandomInt(1, this.difficulty * 200)
    let secondNumber = super.getRandomInt(1, this.difficulty * 200)

    return [super.toArray(firstNumber), super.toArray(secondNumber)]
  }
  generateOperationPerfect() {
    let digit0 = Math.max(1, Math.round(this.difficulty * 4));
    let firstNumber = []
    let secondNumber = null
    while (secondNumber === null) {
      firstNumber[0] = super.getRandomInt(1, 9)
      secondNumber = null
      for (let a = 0; a < digit0; a++) {
        firstNumber[a] = super.getRandomDigit();
      }
      secondNumber = this.createSecondNumber(firstNumber)
      console.log(secondNumber)
      if (secondNumber !== null) {
        console.log(super.toArray(secondNumber))
      }
    }
    return [firstNumber, super.toArray(secondNumber)]

  }
  createSecondNumber(firstNumber) {
    let dividendo = super.toNumber(firstNumber);
    console.log(dividendo)
    let factors = super.getAllFactorsFor(dividendo);
    console.log(factors)
    if (factors.length <= 2) {
      return null;
    }
    return factors[super.getRandomInt(0, factors.length - 2)]
  }
  generateOperation() {
    let digit0 = Math.max(1, Math.round(this.difficulty * 4));
    let digit1 = super.getRandomInt(1, Math.max(1, this.difficulty * 2));
    let firstNumber = [super.getRandomInt(1, 9)]
    let secondNumber = [super.getRandomInt(1, 9)]
    for (let a = 0; a < digit0; a++) {
      firstNumber[a] = super.getRandomDigit();
    }
    for (let a = 0; a < digit1; a++) {
      secondNumber[a] = super.getRandomDigit();
    }
    return [firstNumber, secondNumber]
  }
  toArray() {
    let numbers = []
    if (this.mode === 0) {
      numbers = this.generateOperationPerfect();
    } else if (this.mode === 1) {
      numbers = this.generateOperation();
    } else {
      numbers = this.generateFraction();
    }
    let firstNumber = super.toNumber(numbers[0]);
    let secondNumber = super.toNumber(numbers[1]);
    let result = 0
    if (this.mode === 0) {
      result = firstNumber / secondNumber;
    } else if (this.mode === 1) {
      result = Math.round(firstNumber / secondNumber)
    } else {
      result = [firstNumber, secondNumber]
      let gcd = super.gcd(firstNumber, secondNumber)
      while (gcd !== 1) {
        result[0] /= gcd
        result[1] /= gcd
        gcd = super.gcd(result[0], result[1])
      }
    }
    return [firstNumber, secondNumber, result]
  }


}

export { Addition, Subtraction, Division, Multiplication, Pythagoras }

src/Settings.js
import React, { Component } from 'react';

class OperationList extends Component {
  constructor() {
    super();
    this.addOperation = this.addOperation.bind(this);
  }
  operationNumbertoName(number) {
    let string = ''
    switch (number) {
      case 0:
        string = 'Addition'
        break;
      case 1:
        string = 'Subtraction'
        break;
      case 2:
        string = 'Multiplication'
        break;
      case 3:
        string = 'Division'
        break;
      default:
        string = ''
        break;
    }
    return string;
  }
  addOperation(id, mode, amount) {
    this.props.addOperation(id, mode, amount)
  }
  render() {
    let operations = this.props.settings.operation.split(',').map((operation) => {
      return parseInt(operation, 10);
    });
    let self = this;
    console.log(operations)
    return (
      <div className="operationList">
        <ul>
          {operations.map(function (operation, key) {
            let name = self.operationNumbertoName(operation);

            return (
              <li key={key}> <p className="name"> {name} </p> <span className="options">Options</span> <span className="remove">Remove</span> </li>
            )

          })}
          <li className="newOperation">
            <span onClick={(e) => { this.addOperation(0, 0, 10) }}>SUM</span>
            <span onClick={(e) => { this.addOperation(1, 0, 10) }}>SUB</span>
            <span onClick={(e) => { this.addOperation(2, 0, 10) }}>MUL</span>
            <span onClick={(e) => { this.addOperation(3, 0, 10) }}>DIV</span>
          </li>
        </ul>
      </div>
    )
  }
}


class Settings extends Component {
  constructor() {
    super();
    this.addOperation = this.addOperation.bind(this);

  }
  parseUpdateSettings(e, inputSettings) {
    if (inputSettings === {} || inputSettings === undefined) {
      console.error("Nothing was defined for settings, something went wrong.")
      return null;
    }
    console.log(e)
    console.log(inputSettings)
    let settings = inputSettings;
    if (settings.difficulty !== undefined) {

    }
    for (let a in settings) {

    }
  }
  addOperation(id, mode, amount) {
    this.props.addOperation(id, mode, amount);
  }
  render() {

    let difficulty = this.props.difficulty;
    return (
      <div className="settings">
        <div className="difficulty">
          <p className="name">Difficulty</p>
          <ul className="difficultySelector">
            <li onClick={(e) => { this.parseUpdateSettings(e, { difficulty: 0.2 }) }} className={(difficulty === 0.2) ? 'selected' : ''}>Easy</li>
            <li onClick={(e) => { this.parseUpdateSettings(e, { difficulty: 0.5 }) }} className={(difficulty === 0.5) ? 'selected' : ''}>Medium</li>
            <li onClick={(e) => { this.parseUpdateSettings(e, { difficulty: 1 }) }} className={(difficulty === 1) ? 'selected' : ''}>Hard</li>
          </ul>
        </div>
        <div className="operations">
          <p className="name">Exercise</p>
          <ul>
            <li>
              <OperationList parse={this.parseUpdateSettings} addOperation={this.addOperation} settings={this.props.settings} />
            </li>
          </ul>
        </div>
      </div>
    )
  }
}

export default Settings;